server/src/main/java/io/crate/expression/predicate/NotPredicate.java

```diff
@@ -161,8 +179,11 @@ public Void visitFunction(Function function, NullabilityContext context) {
                     return null;
                 }
             }
+            var savedContext = context.childContext();
             for (Symbol arg : function.arguments()) {
-                arg.accept(this, context);
+                var childContext = savedContext.childContext();
+                arg.accept(this, childContext);
```

Comment:
This allows the leaf Reference to see if there is any function that are `NON_NULLABLE` in the path from the top-level  node down the itself.